// @shopware-pwa/nuxt-module/dist/extendNuxtConfig.cjs.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/nuxt-module/src/extendNuxtConfig.ts
var extendNuxtConfig_exports = {};
__export(extendNuxtConfig_exports, {
  default: () => extendNuxtConfig
});
module.exports = __toCommonJS(extendNuxtConfig_exports);
var import_lodash = require("lodash");
function customizer(objValue, srcValue) {
  if ((0, import_lodash.isArray)(objValue)) {
    return objValue.concat(srcValue);
  }
}
var defaultConfig = {
  telemetry: false,
  vue: {
    config: {
      devtools: process.env.ENABLE_DEVTOOLS === "true" || process.env.NODE_ENV !== "production"
    }
  },
  env: {
    CHOKIDAR_USEPOLLING: process.env.NODE_ENV == "production" ? "0" : "1",
    EXPERIMENTAL_IMAGE_PROCESSING_SERVER: process.env.EXPERIMENTAL_IMAGE_PROCESSING_SERVER || "",
    NODE_ENV: process.env.NODE_ENV || "",
    ENABLE_DEVTOOLS: process.env.ENABLE_DEVTOOLS || ""
  },
  server: {
    port: process.env.PORT || 3e3,
    host: process.env.HOST || "0.0.0.0"
  },
  head: {
    title: "Shopware PWA",
    meta: [
      { hid: "project-type", name: "project-type", content: "shopware-pwa" },
      { charset: "utf-8" },
      { name: "viewport", content: "width=device-width, initial-scale=1" },
      { hid: "description", name: "description", content: "" },
      { name: "generator", content: "Vue Storefront 2" }
    ],
    link: [
      { rel: "icon", type: "image/x-icon", href: "/favicon.ico" },
      {
        rel: "preload",
        href: "https://fonts.googleapis.com/css?family=Raleway:300,400,400i,500,600,700%7CRoboto:300,300i,400,400i,500,700&display=swap",
        as: "style"
      },
      {
        rel: "stylesheet",
        href: "https://fonts.googleapis.com/css?family=Raleway:300,400,400i,500,600,700%7CRoboto:300,300i,400,400i,500,700&display=swap",
        media: "print",
        onload: "this.media='all'",
        once: true
      }
    ]
  },
  css: ["~assets/scss/main.scss"],
  plugins: [],
  components: false,
  buildModules: ["@nuxt/typescript-build", "@shopware-pwa/nuxt-module"],
  modules: [
    "@nuxtjs/axios",
    "@nuxtjs/pwa"
  ],
  axios: {},
  pwa: {
    manifest: {
      lang: "en"
    },
    workbox: {
      offlineStrategy: "StaleWhileRevalidate"
    }
  },
  build: {
    babel: {
      plugins: [["@babel/plugin-proposal-private-methods", { loose: true }]]
    }
  },
  render: {
    bundleRenderer: {
      runInNewContext: false
    }
  },
  generate: {
    exclude: [
      /^\/search/
    ]
  }
};
var configs = [defaultConfig];
function extendNuxtConfig(config) {
  configs.push(config);
  return configs.reduce((prev, next) => (0, import_lodash.mergeWith)(prev, next, customizer));
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {});
