// @shopware-pwa/helpers/dist/helpers.cjs.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/helpers/src/index.ts
var src_exports = {};
__export(src_exports, {
  _parseUrlQuery: () => _parseUrlQuery,
  debounce: () => debounce,
  getCategoryUrl: () => getCategoryUrl,
  getCmsEntityByType: () => getCmsEntityByType,
  getCmsLayoutConfiguration: () => getCmsLayoutConfiguration,
  getCmsLink: () => getCmsLink,
  getCmsLinkTarget: () => getCmsLinkTarget,
  getCmsTechnicalPath: () => getCmsTechnicalPath,
  getListingFilters: () => getListingFilters,
  getProductCalculatedListingPrice: () => getProductCalculatedListingPrice,
  getProductCalculatedPrice: () => getProductCalculatedPrice,
  getProductFreeShipping: () => getProductFreeShipping,
  getProductFromPrice: () => getProductFromPrice,
  getProductMainImageUrl: () => getProductMainImageUrl,
  getProductMediaGallery: () => getProductMediaGallery,
  getProductName: () => getProductName,
  getProductNumber: () => getProductNumber,
  getProductOptions: () => getProductOptions,
  getProductPriceDiscount: () => getProductPriceDiscount,
  getProductPriceDiscountPercentage: () => getProductPriceDiscountPercentage,
  getProductProperties: () => getProductProperties,
  getProductQtySteps: () => getProductQtySteps,
  getProductRatingAverage: () => getProductRatingAverage,
  getProductRealPrice: () => getProductRealPrice,
  getProductReviews: () => getProductReviews,
  getProductThumbnailUrl: () => getProductThumbnailUrl,
  getProductTierPrices: () => getProductTierPrices,
  getProductUrl: () => getProductUrl,
  getProductVariantsFromPrice: () => getProductVariantsFromPrice,
  getTranslatedProperty: () => getTranslatedProperty,
  isLinkCategory: () => isLinkCategory,
  isProductSimple: () => isProductSimple,
  loadScript: () => loadScript,
  mapCountries: () => mapCountries,
  mapSalutations: () => mapSalutations,
  toggleSearchFilter: () => toggleSearchFilter
});
module.exports = __toCommonJS(src_exports);

// packages/helpers/src/product/getProductFreeShipping.ts
function getProductFreeShipping(product) {
  return (product == null ? void 0 : product.shippingFree) || false;
}

// packages/helpers/src/product/getProductMainImageUrl.ts
function getProductMainImageUrl(product) {
  var _a, _b, _c;
  return ((_b = (_a = product == null ? void 0 : product.cover) == null ? void 0 : _a.media) == null ? void 0 : _b.url) || ((_c = product == null ? void 0 : product.cover) == null ? void 0 : _c.url) || "";
}

// packages/helpers/src/product/getProductMediaGallery.ts
function getProductMediaGallery({
  product
} = {}) {
  return product && product.media ? product.media.map((media) => {
    const smallThumb = media.media && media.media.thumbnails && media.media.thumbnails.find((thumb) => thumb.width == "400");
    const normalThumb = media.media && media.media.thumbnails && media.media.thumbnails.find((thumb) => thumb.width == "800");
    const bigThumb = media.media && media.media.thumbnails && media.media.thumbnails.find((thumb) => thumb.width == "1920");
    return {
      icon: { url: smallThumb ? smallThumb.url : media.media.url },
      mobile: { url: normalThumb ? normalThumb.url : media.media.url },
      desktop: { url: bigThumb ? bigThumb.url : media.media.url },
      alt: product.name || " "
    };
  }) : [];
}

// packages/helpers/src/product/getProductName.ts
function getProductName({ product } = {}) {
  if (!product) {
    return null;
  }
  return getTranslatedProperty(product, "name");
}

// packages/helpers/src/product/getProductNumber.ts
function getProductNumber(product) {
  return product == null ? void 0 : product.productNumber;
}

// packages/helpers/src/product/getProductOptions.ts
function getProductOptions({
  product
} = {}) {
  var _a;
  return ((_a = product == null ? void 0 : product.options) == null ? void 0 : _a.map((option) => ({
    label: getTranslatedProperty(option, "name"),
    code: option.id,
    value: getTranslatedProperty(option, "name"),
    color: option.colorHexCode
  }))) || [];
}

// packages/helpers/src/product/getProductProperties.ts
function getProductProperties({
  product
} = {}) {
  var _a;
  const propertyList = (_a = product == null ? void 0 : product.properties) == null ? void 0 : _a.map((property) => ({
    name: getTranslatedProperty(property.group, "name"),
    value: getTranslatedProperty(property, "name")
  }));
  return propertyList || [];
}

// packages/helpers/src/product/getProductReviews.ts
function getProductReviews({
  product
} = {}) {
  if (!product || !product.productReviews) {
    return [];
  }
  return product.productReviews.map(({ id, externalUser, customerId, createdAt, content, points }) => ({
    id,
    author: externalUser ? externalUser : customerId,
    date: createdAt,
    message: content,
    rating: points
  }));
}

// packages/helpers/src/product/getProductTierPrices.ts
function getProductTierPrices(product) {
  var _a;
  if (!product || !((_a = product.calculatedPrices) == null ? void 0 : _a.length)) {
    return [];
  }
  const size = product.calculatedPrices.length;
  return product.calculatedPrices.map(({ unitPrice, quantity }, index) => ({
    label: index === size - 1 ? `from ${quantity}` : `to ${quantity}`,
    quantity,
    unitPrice
  }));
}

// packages/helpers/src/product/getProductUrl.ts
function getProductUrl(product) {
  var _a, _b;
  if (!product)
    return "/";
  const seoUrl = (_b = (_a = product.seoUrls) == null ? void 0 : _a[0]) == null ? void 0 : _b.seoPathInfo;
  return seoUrl ? `/${seoUrl}` : `/detail/${product.id}`;
}

// packages/helpers/src/product/getProductRatingAverage.ts
function getProductRatingAverage(product) {
  return product == null ? void 0 : product.ratingAverage;
}

// packages/helpers/src/product/isProductSimple.ts
function isProductSimple({
  product
} = {}) {
  return !!(product == null ? void 0 : product.parentId) && product.id !== product.parentId;
}

// packages/helpers/src/product/getProductThumbnailUrl.ts
function getProductThumbnailUrl(product) {
  var _a, _b, _c, _d, _e;
  const coverImageUrlFallback = ((_b = (_a = product == null ? void 0 : product.cover) == null ? void 0 : _a.media) == null ? void 0 : _b.url) || "";
  const thumbnailImage = ((_e = (_d = (_c = product == null ? void 0 : product.cover) == null ? void 0 : _c.media) == null ? void 0 : _d.thumbnails) == null ? void 0 : _e.length) && product.cover.media.thumbnails.reduce(function(res, thumb) {
    return thumb.width < res.width ? thumb : res;
  }) || null;
  return (thumbnailImage == null ? void 0 : thumbnailImage.url) || coverImageUrlFallback;
}

// packages/helpers/src/product/getProductPriceDiscount.ts
function getProductPriceDiscount(product) {
  var _a, _b;
  return (_b = (_a = product == null ? void 0 : product.calculatedPrice) == null ? void 0 : _a.listPrice) == null ? void 0 : _b.discount;
}

// packages/helpers/src/product/getProductPriceDiscountPercentage.ts
function getProductPriceDiscountPercentage(product) {
  var _a, _b;
  return (_b = (_a = product == null ? void 0 : product.calculatedPrice) == null ? void 0 : _a.listPrice) == null ? void 0 : _b.percentage;
}

// packages/helpers/src/product/getProductCalculatedPrice.ts
function getProductCalculatedPrice(product) {
  var _a;
  return (_a = product == null ? void 0 : product.calculatedPrice) == null ? void 0 : _a.unitPrice;
}

// packages/helpers/src/product/getProductCalculatedListingPrice.ts
function getProductCalculatedListingPrice(product) {
  var _a, _b, _c;
  return ((_b = (_a = product == null ? void 0 : product.calculatedPrice) == null ? void 0 : _a.listPrice) == null ? void 0 : _b.price) || ((_c = product == null ? void 0 : product.calculatedPrice) == null ? void 0 : _c.unitPrice);
}

// packages/helpers/src/product/getProductRealPrice.ts
function getProductRealPrice(product) {
  var _a;
  if (!product) {
    return;
  }
  const real = product.calculatedPrice;
  if (((_a = product.calculatedPrices) == null ? void 0 : _a.length) > 1) {
    return product.calculatedPrices[product.calculatedPrices.length - 1];
  }
  return real;
}

// packages/helpers/src/product/getProductVariantsFromPrice.ts
function getProductVariantsFromPrice(product) {
  if (!product) {
    return;
  }
  const realPrice = getProductRealPrice(product);
  const cheapestPrice = product.calculatedCheapestPrice;
  if ((cheapestPrice == null ? void 0 : cheapestPrice.unitPrice) != (realPrice == null ? void 0 : realPrice.unitPrice)) {
    return cheapestPrice.unitPrice;
  }
}

// packages/helpers/src/product/getProductFromPrice.ts
function getProductFromPrice(product) {
  var _a;
  if (!product) {
    return;
  }
  const realPrice = getProductRealPrice(product);
  const displayFromPriceLabel = ((_a = product.calculatedPrices) == null ? void 0 : _a.length) > 0;
  if (displayFromPriceLabel)
    return realPrice == null ? void 0 : realPrice.unitPrice;
}

// packages/helpers/src/product/getProductQtySteps.ts
function getProductQtySteps(product, availableStockParam = 100, qtySteps = 50) {
  if (!product.purchaseSteps || product.purchaseSteps <= 1)
    return null;
  const availableStock = product.availableStock || availableStockParam;
  const stepsFromStock = Math.floor(availableStock / product.purchaseSteps);
  const steps = qtySteps < stepsFromStock ? qtySteps : stepsFromStock;
  let options = [];
  let i = 1;
  while (i <= steps) {
    options.push(product.purchaseSteps * i++);
  }
  return options;
}

// packages/helpers/src/category/getCategoryUrl.ts
var getCategoryUrl = (category) => {
  var _a, _b;
  if (!category)
    return "/";
  switch (category.type) {
    case "link":
      return getTranslatedProperty(category, "externalLink") || "/";
    case "folder":
      return "/";
    default:
      return ((_b = (_a = category.seoUrls) == null ? void 0 : _a[0]) == null ? void 0 : _b.seoPathInfo) ? `/${category.seoUrls[0].seoPathInfo}` : category.id ? `/navigation/${category.id}` : "/";
  }
};
var isLinkCategory = (category) => (category == null ? void 0 : category.type) === "link";

// packages/helpers/src/country/mapCountries.ts
function mapCountries(countries) {
  if (!(countries == null ? void 0 : countries.length))
    return [];
  const mappedCountries = countries.map((country) => ({
    name: getTranslatedProperty(country, "name"),
    id: country.id
  }));
  return mappedCountries.filter((country) => !!country.name);
}

// packages/helpers/src/cms/getCmsEntityByType.ts
function getCmsEntityByType(page) {
  if (!(page == null ? void 0 : page.resourceType)) {
    return;
  }
  switch (page.resourceType) {
    case "frontend.navigation.page":
      return page.category;
    case "frontend.detail.page":
      return page.product;
  }
}

// packages/helpers/src/cms/getCmsLink.ts
function getCmsLink(content) {
  var _a;
  return ((_a = content == null ? void 0 : content.data) == null ? void 0 : _a.url) || "";
}

// packages/helpers/src/cms/getCmsLinkTarget.ts
function getCmsLinkTarget(content) {
  var _a;
  const inNewTab = (_a = content == null ? void 0 : content.data) == null ? void 0 : _a.newTab;
  return inNewTab ? "_blank" : "_self";
}

// packages/helpers/src/cms/getCmsTechnicalPath.ts
var PRODUCT_PAGE_PREFIX = "/detail";
var NAVIGATION_PAGE_PREFIX = "/navigation";
var getProductTechnicalPath = (product) => `${PRODUCT_PAGE_PREFIX}/${product.id}`;
var getNavigationTechnicalPath = (cmsPage) => `${NAVIGATION_PAGE_PREFIX}/${cmsPage.resourceIdentifier}`;
function getCmsTechnicalPath(page) {
  if (!(page == null ? void 0 : page.resourceType)) {
    return;
  }
  switch (page.resourceType) {
    case "frontend.navigation.page":
      return getNavigationTechnicalPath(page);
    case "frontend.detail.page":
      return getProductTechnicalPath(page.product);
    default:
      throw Error(`Cannot extract a technical URL for provided page type: ${page.resourceType}`);
  }
}

// packages/helpers/src/cms/getCmsLayoutConfiguration.ts
function getCmsLayoutConfiguration(content) {
  if (!content) {
    return {
      cssClasses: null,
      layoutStyles: {}
    };
  }
  return {
    cssClasses: content.cssClass,
    layoutStyles: {
      backgroundColor: content.backgroundColor,
      backgroundImage: content.backgroundMedia ? `url(${content.backgroundMedia.url})` : null,
      marginBottom: content.marginBottom,
      marginLeft: content.marginLeft,
      marginRight: content.marginRight,
      marginTop: content.marginTop
    }
  };
}

// packages/helpers/src/urlHelpers.ts
function _parseUrlQuery(query) {
  const searchCriteria = {};
  if (!query || typeof query !== "object") {
    return searchCriteria;
  }
  Object.keys(query).forEach((key) => {
    try {
      searchCriteria[key] = typeof query[key] === "string" && ["{", "["].includes(query[key].charAt(0)) ? JSON.parse(query[key]) : query[key];
    } catch (e) {
      console.error("[helpers][parseUrlQuery] Problem with resolving url param: " + key);
    }
  });
  return searchCriteria;
}

// packages/helpers/src/listing/filters.ts
function toggleSearchFilter(currentFilters, filter) {
  let resultFilterValue = void 0;
  if (["range", "max"].includes(filter.type || "")) {
    if (filter.value)
      resultFilterValue = filter.value;
  } else {
    let filterCopy = currentFilters[filter.code] || [];
    if (!Array.isArray(filterCopy))
      filterCopy = [filterCopy];
    const index = filterCopy.findIndex((element) => element === filter.value);
    if (index < 0) {
      filterCopy.push(filter.value);
    } else {
      filterCopy.splice(index, 1);
    }
    if (filterCopy.length)
      resultFilterValue = filterCopy;
  }
  return Object.assign({}, __spreadProps(__spreadValues({}, currentFilters), {
    [filter.code]: resultFilterValue
  }));
}

// packages/helpers/src/listing/getListingFilters.ts
var getFilter = (code, aggregation) => {
  return __spreadValues({
    label: getTranslatedProperty(aggregation, "name") || code,
    code
  }, aggregation);
};
function getListingFilters(aggregations) {
  if (!aggregations) {
    return [];
  }
  const transformedFilters = [];
  for (const [aggregationName, aggregation] of Object.entries(aggregations)) {
    if (aggregationName === "properties" && aggregation.entities) {
      for (const property of aggregation.entities) {
        transformedFilters.push(getFilter(aggregationName, property));
      }
    } else if (!["properties", "options"].includes(aggregationName)) {
      transformedFilters.push(getFilter(aggregationName, aggregation));
    }
  }
  return transformedFilters;
}

// packages/helpers/src/salutation/mapSalutations.ts
function mapSalutations(salutations) {
  if (!(salutations == null ? void 0 : salutations.length))
    return [];
  const mappedSalutations = salutations.map((salutation) => {
    var _a;
    return {
      name: (_a = salutation.displayName) != null ? _a : salutation.salutationKey,
      id: salutation.id
    };
  });
  return mappedSalutations.filter((mappedSalutation) => mappedSalutation.name !== null);
}

// packages/helpers/src/plugins/loadScript.ts
function loadScript(src) {
  return new Promise(function(resolve, reject) {
    if (!document || document.querySelector('script[src="' + src + '"]')) {
      resolve();
      return;
    }
    const el = document.createElement("script");
    el.type = "text/javascript";
    el.async = true;
    el.src = src;
    el.addEventListener("load", resolve);
    el.addEventListener("error", reject);
    el.addEventListener("abort", reject);
    document.head.appendChild(el);
  });
}

// packages/helpers/src/debounce.ts
function debounce(fn, delay = 300) {
  let prevTimer = null;
  return (...args) => {
    if (prevTimer) {
      clearTimeout(prevTimer);
    }
    prevTimer = window.setTimeout(() => {
      fn(...args);
      prevTimer = null;
    }, delay);
  };
}

// packages/helpers/src/getTranslatedProperty.ts
function getTranslatedProperty(element, property) {
  var _a;
  return ((_a = element == null ? void 0 : element.translated) == null ? void 0 : _a[property]) || (element == null ? void 0 : element[property]) || "";
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  _parseUrlQuery,
  debounce,
  getCategoryUrl,
  getCmsEntityByType,
  getCmsLayoutConfiguration,
  getCmsLink,
  getCmsLinkTarget,
  getCmsTechnicalPath,
  getListingFilters,
  getProductCalculatedListingPrice,
  getProductCalculatedPrice,
  getProductFreeShipping,
  getProductFromPrice,
  getProductMainImageUrl,
  getProductMediaGallery,
  getProductName,
  getProductNumber,
  getProductOptions,
  getProductPriceDiscount,
  getProductPriceDiscountPercentage,
  getProductProperties,
  getProductQtySteps,
  getProductRatingAverage,
  getProductRealPrice,
  getProductReviews,
  getProductThumbnailUrl,
  getProductTierPrices,
  getProductUrl,
  getProductVariantsFromPrice,
  getTranslatedProperty,
  isLinkCategory,
  isProductSimple,
  loadScript,
  mapCountries,
  mapSalutations,
  toggleSearchFilter
});
